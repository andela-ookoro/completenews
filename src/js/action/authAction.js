
import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

/**
 * dispatches the user's current state
 * @param {boolean} status - The user's current authentication value
 * @param {object} userinfo - An object that content the user name, email and
 * profile image link and favourite articles count
 * @return {null } Return nothing.
*/
const authAction = (status, userinfo) => {
  if (typeof status === 'boolean') {
    Dispatcher.dispatch({
      Type: Constant.GET_AUTH_STATUS,
      status,
      userinfo,
    });
  } else {
    Dispatcher.dispatch({
      Type: Constant.GET_NOTIFY,
      message: 'Invalid datatype',
    });
  }
};

export default authAction;
