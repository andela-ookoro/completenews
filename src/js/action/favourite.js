
import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

const GETFAVOURITE = (i = 0) => {
  let favouriteCount = JSON.stringify(localStorage.getItem('favourite'));
  const favouriteArray = favouriteCount.split(']');
  favouriteCount = favouriteArray.length + i;
  Dispatcher.dispatch({
    Type: Constant.GETFAVOURITE_COUNT,
    favouriteCount,
  });
};

export default GETFAVOURITE;
