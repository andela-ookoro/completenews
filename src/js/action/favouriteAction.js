
import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

/**
 * dispatches the user favourite article count
 * @param {int} i - An option array; which would be added to the favourite
 * articles count
 * @return {object } Return nothing.
*/
const favouriteAction = (i = 0) => {
  // check if variable exist
  let favourites = [];
  if (localStorage.getItem('favoutireArticles')) {
    favourites = JSON.parse(localStorage.getItem('favoutireArticles'));
  }

  if (i === 0) {
    const favouriteCount = favourites.length;
    Dispatcher.dispatch({
      Type: Constant.GET_FAVOURITE_COUNT,
      favouriteCount,
      countchange: false
    });
  } else {
    Dispatcher.dispatch({
      Type: Constant.GET_FAVOURITE_COUNT,
      favouriteCount: i,
      countchange: true
    });
  }
};
export default favouriteAction;
