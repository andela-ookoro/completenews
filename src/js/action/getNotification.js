import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

/**
 * dispatch a message
 * @param {string} message - the message that should be dispatched
 * @return {string } Return nothing.
*/
const getNotification = (message) => {
  Dispatcher.dispatch({
    Type: Constant.GET_NOTIFY,
    message,
  });
};

export default getNotification;
