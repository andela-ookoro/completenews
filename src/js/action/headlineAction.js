import Dispatcher from '../dispatcher/Dispatcher';
import * as Api from '../utilities/api';
import * as Constant from '../constants';

/**
 * get articles from the api and dispatch them
 *  @param {string} source - the source to of the articles
 * @param {string} sort - the sort parameter
 * @return {string } Return nothing.
*/
export const getArticles = (source, sort = '') => {
  Api.getArticles(source, sort)
    .then((articles) => {
      Dispatcher.dispatch({
        Type: Constant.GET_ARTICLES,
        articles,
        source,
      });
    })
    .catch((err) => {
      Dispatcher.dispatch({
        Type: Constant.GET_NOTIFY,
        message: err,
      });
    });
};

/**
 * get users favourite articles from the database  and dispatch them
 *  @param {string} userEmail - the user email
 * @return {string } Return nothing.
*/
export const getFavouriteArticles = (userEmail) => {
  Api.getFavouriteArticles(userEmail)
    .then((articles) => {
      Dispatcher.dispatch({
        Type: Constant.GET_FAVOURITE_ARTICLES,
        articles,
      });
    })
    .catch((err) => {
      Dispatcher.dispatch({
        Type: Constant.GET_NOTIFY,
        message: err,
      });
    });
};

/**
 * reset the articles to an empty array
 * @return {string } Return nothing.
*/
export const resetArticles = () => {
  Dispatcher.dispatch({
    Type: Constant.GET_ARTICLES,
    articles: [],
  });
};
