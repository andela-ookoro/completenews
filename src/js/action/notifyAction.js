import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

/**
 * dispatch a messah=ge
 * @param {string} message - the message that should be dispatched
 * @return {string } Return nothing.
*/
const getNotification = (message) => {
  Dispatcher.dispatch({
    Type: Constant.GET_NOTIFY,
    message,
  });
};

export default getNotification;
