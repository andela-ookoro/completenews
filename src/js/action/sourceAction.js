import Dispatcher from '../dispatcher/Dispatcher';
import * as Api from '../utilities/api';

export const getSources = () => {
  Api.getSources((sources) => {
    Dispatcher.dispatch({
      actionType: 'GET-SOURCES',
      sources,
    });
  });
};
