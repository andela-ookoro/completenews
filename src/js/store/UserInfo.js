import { EventEmitter } from 'events';
import Dispatcher from '../dispatcher/Dispatcher';
import * as Constant from '../constants';

/**
 * @FileOverview A class that listen to action dispatched for Authentication
 * status
 * and emit a change.
 * @Author okwudiri.okoro@andela.com (Okoro Celestine)
 */
class UserInfo extends EventEmitter {
  /** Create UserInfo object. */
  constructor() {
    super();
    this.updateUserInfo = this.updateUserInfo.bind(this);
    this.isAuth = false;
    this.userinfo = {};
  }

  /**
   * Set the sources .
   * @param {boolean} status - The authentication status to be set
   * @param {boolean} userinfo - The user metadata status to be set
   * @return {boolean} The authentication status .
  */
  updateUserInfo(status, userinfo) {
    this.isAuth = status;
    this.userinfo = userinfo;
    return this.isAuth;
  }

  /**
   * Listen to action dispatched and react to action of type GET_AUTH_STATUS
   * @param {object} action - The an action that was dispatched
   * @return {null} It return no value
  */
  handleActions(action) {
    switch (action.Type) {
      case Constant.GET_AUTH_STATUS:
        this.updateUserInfo(action.status, action.userinfo);
        this.emit('change');
        break;
      default :
    }
  }
}

const NewUserInfo = new UserInfo();
Dispatcher.register(NewUserInfo.handleActions.bind(NewUserInfo));
export default NewUserInfo;

