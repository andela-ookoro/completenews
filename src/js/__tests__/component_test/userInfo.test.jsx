import React from 'react';
import { mount } from 'enzyme';
import Renderer from 'react-test-renderer';
import UserInfo from '../../pages/userinfo';
import * as ArticlesAction from '../../action/articleAction';
import Mockdata from '../../__mocks__/mockData';

describe('userinfo component', () => {
  it('snapshot should be same as render content', () => {
    const component = Renderer.create(<UserInfo />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  const wrapper1 = mount(<UserInfo />);

  it('Should have an initial state with; an empty userInfo object, ' +
    ' isAuth to be false and favourite count to be one',
    () => {
      const state = wrapper1.node.state;
      expect(state.UserInfo).toEqual({});
      expect(state.isAuth).toBe(false);
    });

  localStorage.setItem('userProfile', JSON.stringify(Mockdata.userInfo));
  localStorage.setItem('articles', JSON.stringify(Mockdata.articles));
  const wrapper2 = mount(<UserInfo />);
  const state1 = wrapper2.node.state;
  it('Should display the user image, name; a signout button ' +
    'and favourite button for authenticated user using data in local storage',
    () => {
      expect(state1.UserInfo).toEqual(Mockdata.userInfo);
      expect(state1.isAuth).toBe(true);
    });

  ArticlesAction.getFavouriteArticles = jest.fn(email =>
    `mock userinfo ${email}`
  );
  const newUserInfo = new UserInfo();
  it('should have a function \'viewFavourite\' that invokes ' +
    'the \'getFavouriteArticles\' function in  article action',
  () => {
    expect(newUserInfo.viewFavourite).toBeInstanceOf(Function);
    newUserInfo.viewFavourite('okorocelestine');
    expect(ArticlesAction.getFavouriteArticles).toBeCalledWith('okorocelestine');
  });

  it('should have a function; \'signout\' that resets the user\'s info' +
    ' on localstorage and the component state ',
  () => {
    expect(newUserInfo.signout).toBeInstanceOf(Function);
    newUserInfo.signout();
    const state = newUserInfo.state;
    expect(state.UserInfo).toEqual({});
    expect(state.isAuth).toBe(false);
    expect(state.favouriteCount).toBe(0);
  });
});
