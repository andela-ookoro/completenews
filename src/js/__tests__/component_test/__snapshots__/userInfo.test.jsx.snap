// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userinfo component snapshot should be same as render content 1`] = `
<div>
  <button
    id="viewfavouritebtn"
    onClick={[Function]}
    title="view favourite headline"
  >
     Favourite Articles (
    0
    )
  </button>
  <a
    className="btn-flat"
    onClick={[Function]}
  >
    Sign Out
  </a>
  <div
    className="chip"
  >
    <img
      alt="Contact Person"
      src="test.jpg"
    />
    celestine
  </div>
</div>
`;
